staging/src/k8s.io/apimachinery/pkg/util/waitgroup/ratelimited_waitgroup_test.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package waitgroup
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"golang.org/x/time/rate"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func TestRateLimitedSafeWaitGroup(t *testing.T) {
+	// we want to keep track of how many times rate limiter Wait method is
+	// being invoked, both before and after the wait group is in waiting mode.
+	limiter := &limiterWrapper{}
+
+	// we expect the context passed by the factory to be used
+	var cancelInvoked int
+	factory := &factory{limiter: limiter, grace: 2 * time.Second, ctx: context.Background(), cancel: func() {
+		cancelInvoked++
+	}}
+	target := RateLimitedSafeWaitGroup{limiter: limiter}
+
+	n1, n2 := 100, 101
+	wg1, wg2Add, wg2Done := sync.WaitGroup{}, sync.WaitGroup{}, sync.WaitGroup{}
```

Comment:
Can you add some explanatory comment of what the tests is doing (and how to interpret those waitgroups?)

I think I figured it out, but it would be better if I wouldn't have to reverse-engineer it.

Reply:
yes, added some comments to make it more clear